sphinx/screens/chats/chat-common/chat-common/src/main/java/chat/sphinx/chat_common/util/AudioRecorderController.kt

```diff
@@ -0,0 +1,90 @@
+package chat.sphinx.chat_common.util
+
+import android.media.MediaRecorder
+import androidx.navigation.NavArgs
+import chat.sphinx.chat_common.ui.ChatViewModel
+import chat.sphinx.chat_common.ui.viewstate.footer.FooterViewState
+import io.matthewnelson.concept_coroutines.CoroutineDispatchers
+import io.matthewnelson.concept_media_cache.MediaCacheHandler
+import kotlinx.coroutines.*
+import java.io.File
+
+class AudioRecorderController<ARGS : NavArgs>(
+    private val viewModel: ChatViewModel<ARGS>,
+    private val viewModelScope: CoroutineScope,
+    val mediaCacheHandler: MediaCacheHandler,
+    dispatchers: CoroutineDispatchers,
+): CoroutineDispatchers by dispatchers {
+    var recordingTempFile: File? = null
+    private var recorder: MediaRecorder? = null
+
+    private var dispatchStateJob: Job? = null
+    fun startAudioRecording() {
+        if (dispatchStateJob?.isActive == true) {
+            return
+        }
+
+        recorder = MediaRecorder().apply {
+            recordingTempFile  = mediaCacheHandler.createAudioFile(AUDIO_FORMAT_EXTENSION)
+
+            setAudioSource(MediaRecorder.AudioSource.MIC)
+            setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
+            setOutputFile(recordingTempFile?.absolutePath)
+            setAudioEncoder(MediaRecorder.AudioEncoder.HE_AAC)
+
+            prepare()
+
+            start()
+
+            dispatchStateJob = viewModelScope.launch(mainImmediate) {
+                var duration = 0L
+                while (isActive) {
+                    viewModel.updateFooterViewState(
+                        FooterViewState.RecordingAudioAttachment(duration)
+                    )
+                    if (!isRecording()) {
+                        break
+                    } else {
+                        duration += 250L
+                        delay(250L)
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Reliance on the `dispatchStateJob` running to return if active in this particular use case is fragile, as time between preparation of the recorder and the launching of dispatchStateJob can result in multiple recorder jobs being launched at ones.

This needs to utilize locks.

I'm definitely **for** creation + preparation of a new `MediaRecorder` upon every method call as to not hold on to those resources unnecessarily, but we should be doing so carefully as preparing multiple and setting a single reference variable means we're only controlling a single instance and allowing the others to run freely.